src/main/java/slimeknights/tconstruct/smeltery/recipe/AbstractCastingRecipe.java

```diff
@@ -0,0 +1,202 @@
+package slimeknights.tconstruct.smeltery.recipe;
+
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonSyntaxException;
+import net.minecraft.fluid.Fluid;
+import net.minecraft.fluid.Fluids;
+import net.minecraft.inventory.IInventory;
+import net.minecraft.item.ItemStack;
+import net.minecraft.item.crafting.IRecipe;
+import net.minecraft.item.crafting.IRecipeSerializer;
+import net.minecraft.item.crafting.IRecipeType;
+import net.minecraft.item.crafting.Ingredient;
+import net.minecraft.network.PacketBuffer;
+import net.minecraft.util.JSONUtils;
+import net.minecraft.util.ResourceLocation;
+import net.minecraft.world.World;
+import net.minecraftforge.fluids.FluidStack;
+import net.minecraftforge.registries.ForgeRegistries;
+import net.minecraftforge.registries.ForgeRegistryEntry;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+public abstract class AbstractCastingRecipe implements IRecipe<IInventory> {
```

Comment:
Use a custom interface instead of IInventory. Will let you include fluid in the interface and removes the need for two separate matches methods. The whole design of vanilla recipes is you can choose the container type, so we can add more arguments later.